### PR TITLE
Calling update() no longer required after model's deserialization

### PR DESCRIPTION
**Pull request by cezary986**
Calling the ruleset's `update` method is no longer required after deserializing the model. Prior to these changes, the user had to manually call update on each deserialized ruleset object, passing it the full training dataset. Otherwise, the model was in an incomplete state, and calling some of its methods caused exceptions.

In this PR, I introduced two available serialization modes: 
* `full` - serializes all model information, so it is ready for use after deserialization, no need to call the update method;
* `minimal` - serializes only minimal key information, after deserialization the user must call the update method (works the same as before these changes);

The default serialization mode is `full`, so by default no update is needed after deserialization. The mode can be configured using the `decision_rules.serialization.JSONSerializer.serialize` method, example below:
```python
# using enum (recommended)
from decision_rules.serialization import SerializationModes

JSONSerializer.serialize(ruleset, mode=SerializationModes.FULL)
JSONSerializer.serialize(ruleset, mode=SerializationModes.MINIMAL)

# using the mode name
JSONSerializer.serialize(ruleset, mode='full')
JSONSerializer.serialize(ruleset, mode='minimal')

# will use the default mode “full”
JSONSerializer.serialize(ruleset)
```

It is important to say that the “full” mode can significantly increase the resulting size of json files. For classification and regression rule sets, the difference may be negligible, but for survival rule sets, the size may even double.

### CODE REVIEW

File: decision_rules/core/ruleset.py [hunk 1]
```diff
@@ -52,22 +40,26 @@ def __init__(
...
-        if not self._voting_weights_calculated:
+    def _validate_object_state_before_prediction(self):
+        voting_weights_calculated: bool = all(
+            [rule.coverage is not None for rule in self.rules]
+        )
+        if not voting_weights_calculated:
             raise InvalidStateError(
-                'Rules coverages must have been calculated before prediction.' +
-                'Did you forget to call update(...) method?'
+                "Rules coverages must have been calculated before prediction."
```
Comment (adamgrzelak) [inline decision_rules/core/ruleset.py:55]: To w zasadzie językowa uwaga. "Must have been" sugeruje inne znaczenie, niż chcemy. Zmieniłbym to na:
"Rule coverages have to be calculated before performing prediction."

File: decision_rules/core/ruleset.py [hunk 2]
```diff
@@ -78,16 +70,18 @@ def set_default_conclusion_enabled(self, enabled: bool) -> None:
         """
         if self._stored_default_conclusion is None:
             raise InvalidStateError(
-                'Default conclusion was never configured.' +
-                'Call update(...) method to calculated automatic default conclusion ' +
-                'or set it manually.'
+                "Default conclusion was never configured."
+                + "Call update(...) method to calculated automatic default conclusion "
```
Comment (adamgrzelak) [inline decision_rules/core/ruleset.py:74]: Jest "calculated", powinno być "calculate".

**Review by adamgrzelak — CHANGES_REQUESTED** [review]

File: decision_rules/core/ruleset.py (hunk 2 above)
Comment (cezary986) [reply decision_rules/core/ruleset.py]: [fix: Code review requested changes](https://github.com/ruleminer/decision-rules/pull/15/commits/78eb3b39e6c5a3e80c96b8e8dab81874bd5bc49a)

File: decision_rules/core/ruleset.py (hunk 1 above)
Comment (cezary986) [reply decision_rules/core/ruleset.py]: [fix: Code review requested changes](https://github.com/ruleminer/decision-rules/pull/15/commits/78eb3b39e6c5a3e80c96b8e8dab81874bd5bc49a)

**Review by adamgrzelak — APPROVED** [review]

**Pull request merged**